File: android/core/protocol/src/main/kotlin/app/hapi/protocol/chat/CodexCommandPresentation.kt
package app.hapi.protocol.chat

import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement

/** Port of `web/src/chat/codexCommandPresentation.ts` (grouping eligibility inputs). */

sealed class CodexCommandAction {
    abstract val command: String

    data class Read(override val command: String, val name: String, val path: String) : CodexCommandAction()
    data class ListFiles(override val command: String, val path: String?) : CodexCommandAction()
    data class Search(override val command: String, val query: String?, val path: String?) : CodexCommandAction()
    data class Unknown(override val command: String) : CodexCommandAction()
}

/** TS local `asString`: non-empty strings only. */
private fun nonEmptyString(value: JsonElement?): String? =
    asString(value)?.takeIf { it.isNotEmpty() }

private fun parseAction(value: JsonElement?): CodexCommandAction? {
    val action = asObject(value) ?: return null
    val type = nonEmptyString(action["type"]) ?: return null
    val command = nonEmptyString(action["command"]) ?: return null

    return when (type) {
        "read" -> {
            val name = nonEmptyString(action["name"])
            val path = nonEmptyString(action["path"])
            if (name != null && path != null) CodexCommandAction.Read(command, name, path) else null
        }
        "listFiles" -> CodexCommandAction.ListFiles(command, nonEmptyString(action["path"]))
        "search" -> CodexCommandAction.Search(command, nonEmptyString(action["query"]), nonEmptyString(action["path"]))
        "unknown" -> CodexCommandAction.Unknown(command)
        else -> null
    }
}

fun getCodexCommandActions(block: ToolCallBlock): List<CodexCommandAction> {
    if (block.tool.name != "CodexBash") return emptyList()
    val input = asObject(block.tool.input) ?: return emptyList()
    val raw = input["command_actions"].orNull() ?: input["commandActions"]
    val array = raw as? JsonArray ?: return emptyList()
    return array.mapNotNull(::parseAction)
}

fun isCodexExplorationTool(block: ToolCallBlock): Boolean {
    val input = asObject(block.tool.input)
    val source = nonEmptyString(input?.get("command_source").orNull() ?: input?.get("commandSource"))
    if (source?.lowercase() == "usershell") return false

    val actions = getCodexCommandActions(block)
    return actions.isNotEmpty() && actions.all {
        it is CodexCommandAction.Read || it is CodexCommandAction.ListFiles || it is CodexCommandAction.Search
    }
}

File: web/src/chat/codexCommandPresentation.ts
import type { ToolCallBlock } from '@/chat/types'

export type CodexCommandAction =
    | { type: 'read'; command: string; name: string; path: string }
    | { type: 'listFiles'; command: string; path: string | null }
    | { type: 'search'; command: string; query: string | null; path: string | null }
    | { type: 'unknown'; command: string }

function asString(value: unknown): string | null {
    return typeof value === 'string' && value.length > 0 ? value : null
}

function parseAction(value: unknown): CodexCommandAction | null {
    if (!value || typeof value !== 'object') return null
    const action = value as Record<string, unknown>
    const type = asString(action.type)
    const command = asString(action.command)
    if (!type || !command) return null

    if (type === 'read') {
        const name = asString(action.name)
        const path = asString(action.path)
        return name && path ? { type, command, name, path } : null
    }
    if (type === 'listFiles') {
        return { type, command, path: asString(action.path) }
    }
    if (type === 'search') {
        return {
            type,
            command,
            query: asString(action.query),
            path: asString(action.path)
        }
    }
    if (type === 'unknown') {
        return { type, command }
    }
    return null
}

export function getCodexCommandActions(block: ToolCallBlock): CodexCommandAction[] {
    if (block.tool.name !== 'CodexBash' || !block.tool.input || typeof block.tool.input !== 'object') {
        return []
    }
    const input = block.tool.input as Record<string, unknown>
    const raw = input.command_actions ?? input.commandActions
    if (!Array.isArray(raw)) return []
    return raw.map(parseAction).filter((action): action is CodexCommandAction => action !== null)
}

export function isCodexExplorationTool(block: ToolCallBlock): boolean {
    const input = block.tool.input && typeof block.tool.input === 'object'
        ? block.tool.input as Record<string, unknown>
        : null
    const source = asString(input?.command_source ?? input?.commandSource)
    if (source?.toLowerCase() === 'usershell') return false

    const actions = getCodexCommandActions(block)
    return actions.length > 0 && actions.every((action) => (
        action.type === 'read' || action.type === 'listFiles' || action.type === 'search'
    ))
}
